bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java

```diff
@@ -403,39 +405,53 @@ private Void doConcat(SegmentHandle targetHandle, long offset, String sourceSegm
-        long objectSize = metadataResult.getContentLength(); // in bytes
-
-        copyRequest = new CopyPartRequest(config.getBucket(),
-                config.getPrefix() + sourceSegment,
-                config.getBucket(),
-                targetPath,
-                uploadId,
-                2).withSourceRange(Range.fromOffsetLength(0, objectSize));
-
-        copyResult = client.copyPart(copyRequest);
-        partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
-
-        //Close the upload
-        client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
-                targetPath, uploadId).withParts(partEtags));
-
-        client.deleteObject(config.getBucket(), config.getPrefix() + sourceSegment);
-        Duration elapsed = timer.getElapsed();
+        if (config.getSmallObjectThreshold() < si.getLength()) {
+
+            //Copy the first part
+            CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                    targetPath,
+                    config.getBucket(),
+                    targetPath,
+                    uploadId,
+                    1).withSourceRange(Range.fromOffsetLength(0, offset));
+            CopyPartResult copyResult = client.copyPart(copyRequest);
+
+            partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+
+            //Copy the second part
+            S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getBucket(),
+                    config.getPrefix() + sourceSegment);
+            long objectSize = metadataResult.getContentLength(); // in bytes
+
+            copyRequest = new CopyPartRequest(config.getBucket(),
+                    config.getPrefix() + sourceSegment,
+                    config.getBucket(),
+                    targetPath,
+                    uploadId,
+                    2).withSourceRange(Range.fromOffsetLength(0, objectSize));
+
+            copyResult = client.copyPart(copyRequest);
+            partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
 
+            //Close the upload
+            client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
+                    targetPath, uploadId).withParts(partEtags));
+
+            client.deleteObject(config.getBucket(), config.getPrefix() + sourceSegment);
+
+        } else {
+            try (InputStream reader = client.readObjectStream(config.getBucket(),
+                    config.getPrefix() + sourceSegment, Range.fromOffsetLength(0, si.getLength()))) {
+                client.putObject(this.config.getBucket(),
+                        targetPath,
+                        Range.fromOffsetLength(offset, si.getLength()),
+                        new BufferedInputStream(reader, Math.toIntExact(si.getLength())));
+            } catch (Exception e) {
+                throw Exceptions.sneakyThrow(e);
```

Comment:
Please give more information here, particularly this is something new we're trying

Reply:
actually we don't need to catch exception here.